### PR TITLE
Add static FP8 expert support 

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?

This PR enables static FP8 experts. This also works on multi-gpu with device-map. A fix for that was to set was to set `torch.cuda.set_device()`.  Triton's JIT compiler uses he active device context to determine which GPU to compile and launch the kernel on. If tensors are on  GPU:1 but the active context is still GPU:0, Triton compiles/launches the kernel on the wrong GPU — reading garbage memory from a different device.

I will do a follow-up PR in kernels to have better per-tensor kernels. This will also enable to remove the padding trick that we currently have + add static activation scale for batched and grouped implementation. However, this is not needed for this PR.

### CODE REVIEW

**Comment by SunMarc** [comment]
@bot /style

**Review by ArthurZucker — APPROVED** [review]
LGTM ty, a test would be welcome!

**Comment by SunMarc** [comment]
> LGTM ty, a test would be welcome!

I'll do a much stronger test suite for fp8 in a follow-up PR ! For now, let's merge

File: src/transformers/integrations/finegrained_fp8.py [hunk 1]
```diff
@@ -186,7 +186,29 @@ def w8a8_fp8_matmul(
     if _supports_cutlass(A, B, block_size, output_dtype):
         return w8a8_block_fp8_matmul_cutlass(A, B, As, Bs, output_dtype)
 
-    # Fall back to Triton
+    # Ensure correct CUDA device context for Triton JIT on multi-GPU
+    torch.cuda.set_device(A.device)
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/finegrained_fp8.py:190]: imo this should be in the kernel not outside of it

File: src/transformers/integrations/finegrained_fp8.py [hunk 2]
```diff
@@ -457,6 +478,10 @@ def __init__(
         )
         self.register_parameter("down_proj_bias", None)
 
+        if self.activation_scheme == "static":
+            self.gate_up_proj_activation_scale = nn.Parameter(torch.ones(self.num_experts, dtype=torch.float32))
+            self.down_proj_activation_scale = nn.Parameter(torch.ones(self.num_experts, dtype=torch.float32))
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/finegrained_fp8.py:483]: what models use static act scheme experts ?

**Review by IlyasMoutawwakil — COMMENTED** [review]
sorry for the late review ! thanks for taking care of this 🙏

File: src/transformers/integrations/finegrained_fp8.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/integrations/finegrained_fp8.py:190]: yes, i'll update the kernels to clean this integration later on !

File: src/transformers/integrations/finegrained_fp8.py (hunk 2 above)
Comment (SunMarc) [reply src/transformers/integrations/finegrained_fp8.py:483]: mistral small 4 ! it was added very recently
